routers/user/auth.go

```diff
@@ -1185,3 +1186,69 @@ func ResetPasswdPost(ctx *context.Context) {
 	ctx.Data["IsResetFailed"] = true
 	ctx.HTML(200, tplResetPassword)
 }
+
+// MustChangePassword renders the page to change a user's password
+func MustChangePassword(ctx *context.Context) {
+	ctx.Data["Title"] = ctx.Tr("auth.must_change_password")
+	ctx.Data["ChangePasscodeLink"] = setting.AppSubURL + "/user/settings/change_password"
+
+	ctx.HTML(200, tplMustChangePassword)
+}
+
+// MustChangePasswordPost response for updating a user's password after his/her
+// account was created by an admin
+func MustChangePasswordPost(ctx *context.Context, cpt *captcha.Captcha, form auth.MustChangePasswordForm) {
+	ctx.Data["Title"] = ctx.Tr("auth.must_change_password")
+
+	ctx.Data["ChangePasscodeLink"] = setting.AppSubURL + "/user/settings/change_password"
+
+	if ctx.HasError() {
+		ctx.HTML(200, tplMustChangePassword)
+		return
+	}
+
+	u := ctx.User
+
+	// Make sure only requests for users who are eligible to change their password via
+	// this method passes through
+	if !u.MustChangePassword {
+		ctx.ServerError("MustUpdatePassword", errors.New("cannot update password.. Please visit the settings page"))
+		return
+	}
+
+	if form.Password != form.Retype {
+		ctx.Data["Err_Password"] = true
+		ctx.RenderWithErr(ctx.Tr("form.password_not_match"), tplMustChangePassword, &form)
+		return
+	}
+
+	if len(form.Password) < setting.MinPasswordLength {
+		ctx.Data["Err_Password"] = true
+		ctx.RenderWithErr(ctx.Tr("auth.password_too_short", setting.MinPasswordLength), tplMustChangePassword, &form)
+		return
+	}
+
+	var err error
+	if u.Rands, err = models.GetUserSalt(); err != nil {
```

Comment:
I don't think this should be updated when changing password, only if resetting password.

Reply:
Would fix

Reply:
> User should be allowed to access this password change form only when he has must change password set, otherwise if opening this url manually it will be possible to change password without knowing current password

By the way, the above review has been implemented